Source/Dafny/Verifier/Translator.cs

```diff
@@ -6260,7 +6260,37 @@ private class WFOptions {
 
       } else if (expr is StmtExpr) {
         var e = (StmtExpr)expr;
+        // If we're inside an "old" expression, then "etran" will know how to translate
+        // expressions. However, here, we're also having to translate e.S, which is a
+        // Statement. Since statement translation (in particular, CallStmt's) works
+        // directly on the global variable $Heap, we change its value here.
+        // (Note, this would be a bad translation if the Statement could introduce
+        // control flow that would miss the "$Heap := tmpHeap;" assignment that
+        // resets the value of $Heap. Luckily, StmtExpr does not allow any such
+        // Statement.)
+        Bpl.IdentifierExpr tmpHeap = null;
```

Comment:
`CheckWellformedWithResult` is already very long, so it would be good to extract a method here. 

Consider introducing:
```c#
WithFreshHeap(string heapVarSuffix, Token token, ExpressionTranslator etrans, BoogieStmtListBuilder builder, Action<BoogieStmtListBuilder> addStatements) {
  var suffix = CurrentIdGenerator.FreshId(heapVarSuffix);
  var tmpHeapVar = new Bpl.LocalVariable(token, new Bpl.TypedIdent(token, "Heap$" + suffix, predef.HeapType));
  locals.Add(tmpHeapVar);
  var tmpHeap = new Bpl.IdentifierExpr(token, tmpHeapVar);
  var generalEtran = new ExpressionTranslator(this, predef, token);
  var theHeap = (Bpl.IdentifierExpr /*TODO: this cast is dubious*/)generalEtran.HeapExpr;

  // tmpHeap := $Heap;
  builder.Add(Bpl.Cmd.SimpleAssign(token, tmpHeap, theHeap));
  // $Heap := etran.$Heap;
  builder.Add(Bpl.Cmd.SimpleAssign(token, theHeap, etran.HeapExpr));
  
  addStatements(builder);
  
  // $Heap := tmpHeap;
  builder.Add(Bpl.Cmd.SimpleAssign(token, theHeap, tmpHeap);
}

Reply:
Will do.